dev/devicelab/lib/framework/metrics_center.dart

```diff
@@ -0,0 +1,115 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:metrics_center/metrics_center.dart';
+import 'package:path/path.dart' as p;
+
+Future<ProcessResult> runGit(
+  List<String> args, {
+  String? processWorkingDir,
+}) async {
+  return Process.run(
+    'git',
+    args,
+    workingDirectory: processWorkingDir,
+    runInShell: true,
+  );
+}
+
+/// Returns commit time in committer date in UNIX timestamp.
+Future<String> getCommitDate() async {
+  final String gitRoot = p.absolute('.');
+  final ProcessResult logResult = await runGit(
+    <String>['log', '--pretty=format:%ct', '-n', '1'],
+    processWorkingDir: gitRoot,
+  );
+  if (logResult.exitCode != 0) {
+    throw 'Unexpected exit code ${logResult.exitCode}';
+  }
+  return logResult.stdout.toString().trim();
+}
+
+/// Authenticate and connect to gcloud storage.
+///
+/// It supports both token and credential authentications.
+Future<FlutterDestination> connectFlutterDestination() async {
+  const String kTokenPath = 'TOKEN_PATH';
+  const String kGcpProject = 'GCP_PROJECT';
+  final Map<String, String> env = Platform.environment;
+  final bool isTesting = env['IS_TESTING'] == 'true';
+  if (env.containsKey(kTokenPath) && env.containsKey(kGcpProject)) {
+    return FlutterDestination.makeFromAccessToken(
+      File(env[kTokenPath]!).readAsStringSync(),
+      env[kGcpProject]!,
+      isTesting: isTesting,
+    );
+  }
+  return FlutterDestination.makeFromCredentialsJson(
+    jsonDecode(env['BENCHMARK_GCP_CREDENTIALS']!) as Map<String, dynamic>,
+    isTesting: isTesting,
+  );
+}
+
+/// Parse results into Metric Points.
+///
+/// An example of `resultsJson`:
```

Comment:
To simplify, we can add these to `TaskResult` which will remove the need to pass dynamic json around

Reply:
This logic is executed after task finishes and writes result to the `results-file`, which is the injected value here.

Reply:
It's available - CocoonClient uploads the `TaskResult` https://cs.opensource.google/flutter/flutter/+/master:dev/devicelab/lib/framework/cocoon.dart;l=136

Reply:
It doesn't seem to be an easy change, cocoon is still using existing `updateRequest`, which is different from the properties defined in TaskResult. I prefer to have that change in a separate PR to break anything before validating the effectiveness of the intended logic.
Created https://github.com/flutter/flutter/issues/88483 to track.